src/allotropy/parsers/chemometec_nucleoview/nucleoview_parser.py

```diff
@@ -0,0 +1,173 @@
+    "Dead (cells/ml)": TQuantityValueMillionCellsPerMilliliter,
+    "Estimated cell diameter (um)": TQuantityValueMicrometer,
+    "Live (cells/ml)": TQuantityValueMillionCellsPerMilliliter,
+    "Multiplication factor": TQuantityValueUnitless,
+    "Total (cells/ml)": TQuantityValueMillionCellsPerMilliliter,
+    "Viability (%)": TQuantityValuePercent,
+}
+
+
+def get_property_from_sample(sample: pd.Series, property_name: str) -> Any:
+    if (value := sample.get(property_name)) is not None:
+        property_type = _PROPERTY_LOOKUP[property_name]
+
+        # if the porperty type is measured in million cells per ml convert cells per ml
+        if property_type == TQuantityValueMillionCellsPerMilliliter:
+            return property_type(value=float(str(value)) / 1e6)
+
+        return property_type(value=value)  # type: ignore[arg-type]
+    # special case for cell count since nucleoview doesn't provide total cell count
+    elif property_name == "Cell count":
+        property_type = _PROPERTY_LOOKUP[property_name]
+        return property_type(value=float("NaN"))
+    else:
+        return None
+
+
+class ChemometecNucleoviewParser(VendorParser):
+    def _parse(self, contents: io.IOBase, filename: str) -> Model:
+        return self._get_model(NucleoviewReader.read(contents), filename)
+
+    def _get_model(self, data: pd.DataFrame, filename: str) -> Model:
+        return Model(
+            field_asm_manifest="http://purl.allotrope.org/manifests/cell-counting/BENCHLING/2023/11/cell-counting.manifest",
+            cell_counting_aggregate_document=CellCountingAggregateDocument(
+                device_system_document=DeviceSystemDocument(
+                    model_number=self._get_device_identifier(data),
+                    equipment_serial_number=self._get_device_serial_number(data),
+                ),
+                data_system_document=DataSystemDocument(
+                    file_name=filename,
+                    software_name=NUCLEOCOUNTER_SOFTWARE_NAME,
+                    ASM_converter_name=ASM_CONVERTER_NAME,
+                    ASM_converter_version=ASM_CONVERTER_VERSION,
+                ),
+                cell_counting_document=self._get_cell_counting_document(data),
+            ),
+        )
+
+    def _get_device_serial_number(
+        self,
+        data: pd.DataFrame,
+    ) -> Optional[str]:
+        try:
+            return str(data["Instrument s/n"].iloc[0])
+        except KeyError:
+            return None
+
+    def _get_device_identifier(
+        self,
+        data: pd.DataFrame,
+    ) -> str:
+        try:
+            return str(data["Instrument type"].iloc[0])
+        except KeyError:
+            return DEFAULT_MODEL_NUMBER
+
+    def _get_cell_counting_document(
+        self, data: pd.DataFrame
+    ) -> list[CellCountingDocumentItem]:
+        return [
+            self._get_cell_counting_document_item(sample)
+            for _, sample in data.iterrows()
```

Comment:
oops, I missed this earlier. `iterrows` is a big pandas antipattern. I now know why you did this after looking at other vi-cell code (and fixing it!). Could you please refactor a la #144?

Reply:
latest commit refactors to mimic 144...I think